src/borg/repository.py

```diff
@@ -344,6 +348,28 @@ def write_index(self):
             os.unlink(os.path.join(self.path, name))
         self.index = None
 
+    def check_free_space(self):
+        """Pre-commit check for sufficient free space to actually perform the commit."""
+        # Keep four times the (compact - not necessarily the case) index size free
+        index_size = len(self.index) * 40 * 4
+        required_free_space = index_size
+        if not self.append_only:
+            # Keep one worst-case full free in non-append-only mode
+            required_free_space += self.max_segment_size + MAX_OBJECT_SIZE
+        try:
+            st_vfs = os.statvfs(self.path)
```

Comment:
much easier way to defeat the check is to just run "fallocate" at the right time. there's no way this can be bullet-proof - or even come close to that -, but we can catch the usual case: no strange mount setups, no overlayfs stuff, an FS actually reporting up-to-date numbers, no quotas and (mainly) one writer (our process) to the FS.